Build QualificationRequirements

kitallis opened the pull request:
I'm raising this pull request just so that you can track progress.

It currently supports specifying multiple QRs and passing them along when you're registering a HIT type. It probably needs a few tests before it's merged. I'm not familiar with Midje, so it might take a while until I get around writing those.

I feel this is potentially a bit cluttered in its implementation. See if it makes sense. Meanwhile, I'm also working on handling exceptions and response codes from the MTurk API.

kitallis committed:
adding prismatic schema

kitallis committed:
qualification requirement builder to be supplied when registering a new
HIT type

kitallis committed:
support multiple qualification requirements

kitallis commented:
Here's a usage example:

```
(defn qualification-for-receipt
  []
  (biomass.builder.qualification-requirements/build
   {:qualification-type-id :hits_approved
    :comparator :gte
    :required-to-preview true
    :value 2000}))

(defn hit-type
  []
  (biomass.hits/register-hit-type
   {:title "Transcribe Receipt Details"
    :description "Identify the 1) Biller,  2) Tax"
    :reward-amount 0.10
    :reward-currency "USD"
    :assignment-duration 180} (qualification-for-receipt)))

(defn hit-for-invoice
  [image-url]
  (let [hit-type-id (get-in (hit-type) [:result :hit-type-id])]
    (biomass.hits/create-hit {:hit-type-id hit-type-id
                              :hit-layout-id hit-layout-id
                              :assignment-duration 180
                              :lifetime 43200
                              :layout-params {:image_url image-url}})))

(hit-for-invoice "http://www.sunshinesupport.com/Images/receipt_example.jpg")
```

kitallis committed:
the final param merge needs to be scoped in the loop

kitallis commented:
Largely, things that are pending:
- The three master types have different codes depending on sandbox or production mode, they need to be switched accordingly.
- Needs some testing around passing in multiple QRs.
- Make it work for CreateHIT as well.
- Go through [this](http://docs.aws.amazon.com/AWSMechTurk/latest/AWSMturkAPI/ApiReference_QualificationRequirementDataStructureArticle.html) again and see if anything was unaccounted for.

smnirven commented:
I'll take a look at this tonight -- going to take me a bit to get up to speed on Qualification Requirements.

kitallis committed:
ability to check if we’re making requests in sandbox mode or not

kitallis committed:
change type keys for master categories based on sandbox/production mode

kitallis commented:
@smnirven Any luck?

smnirven commented on src/biomass/builder/qualification_requirements.clj:
```diff
@@ -0,0 +1,92 @@
+
+(defn build
+  [& qualifications]
+  (let [no (+ (count qualifications) 1)]
```
This whole block here smacks of a for loop in other languages. Can't the same thing be accomplished with (map) or something else?

kitallis replied:
Yeah, I think I can just `reduce` over with `map-indexed` instead.

kitallis replied:
```
(defn build
  [& qualifications]
  (reduce merge {} (map-indexed request-params-build qualifcations)))
```

There, with some help from the good folks at #clojure :smile:

smnirven commented on src/biomass/hits.clj:
```diff
@@ -49,17 +49,18 @@
 
 (defn register-hit-type
-  [{:keys [title description reward-amount reward-currency assignment-duration]}]
+  [{:keys [title description reward-amount reward-currency assignment-duration]} qualification-requirements]
```
What if I don't want to add any qualification requirements to a hit-type? Can't qualification-requirements just become another key in the params map, and have the code handle that being null in some cases?

kitallis replied:
It's been a while since I looked at it, but I think the `RegisterHITType` call needs the QRs to be flattened into the map with the rest of the fields, like `title` and `description`.

Which is fine, I think the reason I did that was because QRs are built separately using `biomass.builder.qualification-requirements/build` and as an API you would end up doing a `merge` in your code like:

```
(defn hit-type
  []
  (biomass.hits/register-hit-type
   (merge
    {:title "Transcribe Receipt Details"
     :description "Identify the 1) Biller,  2) Tax"
     :reward-amount 0.10
     :reward-currency "USD"
     :assignment-duration 180} (qualification-for-receipt))))
```

I had also changed the builder to do multiple QRs (afterwards), so in which case you would only merge once. I think it makes sense to do it in the `params` map, yes.

kitallis commented:
Hi, @smnirven. Thanks for the comments. I'll take a look at these tomorrow morning.

smnirven commented:
@kitallis Hey there, let me know if you've got any questions or if I can help!  This is a feature I'd like to add to biomass.

kitallis commented on src/biomass/request.clj:
```diff
@@ -16,13 +16,16 @@
 (def aws-access-key (ref nil))
 (def aws-secret-access-key (ref nil))
 (def base-url (ref nil))
+(def sandbox-mode (ref nil))
```
I don't particularly like this lopsidedness either. Do you think there is a better way to handle this? 

The fact that there's a `sandbox-mode` ref, and then there is `sandboxed?` which just calls it only so that we can access it as a method in https://github.com/smnirven/biomass/pull/6/files#diff-f90f3515814cc6646b0bcedec980a22eR21 seems odd. We could always just use `@sandbox-mode` there, but that didn't read well :weary:

What do you think?

smnirven replied:
Any reason you can't just call the ref sandboxed?

That way you can check it with @sandboxed?

kitallis commented:
Hey @smnirven, I've been loaded with work in the past few weeks. 
I responded to your inline comments. Will try to find some time again this weekend.

Would be great if you could get this PR working for you in whatever state it is, and we can start from there. Because there's a fair bit of permutation going on with the fields you can specify to QRs, the builder really deserves some unit tests, I think.

kitallis committed:
refactor the build function

kitallis commented:
This is still so open. Need to finish this 🐔 /@smnirven

smnirven commented:
@kitallis Correct sir. Should have a chance to look at it this coming weekend, thanks for reminding me!

shafeeq committed:
Multiple changes
- Remove unavailable qualification IDs
- Change method of checking sandboxmode
- Support registering hit without qualification

shafeeq committed:
Change qualifications schema and add tests

shafeeq commented:
Hi @smnirven , I'm building upon @kitallis' changes to implement more operations. You can track my changes here: https://github.com/kitallis/biomass/pull/1

shafeeq committed:
Bump API version

shafeeq committed:
Parse vectors of integers/locales for use with 'In' opertor in qualification requirements

shafeeq committed:
Add some schema, bump dependenceies versions

shafeeq committed:
Register and create hit using new schema and builder

shafeeq committed:
Add schema and functiopns for HIT operations

shafeeq committed:
Add get-assignments-for-hit and approve-rejected-assignment

shafeeq committed:
Convert multiple notifcation types using convert-params

shafeeq committed:
Generic parser for xml responses, change send-and-parse to use it

shafeeq committed:
Refactor code

shafeeq committed:
Add operations based on qualifications

shafeeq committed:
Add worker specific operations

shafeeq committed:
Add misc operations

shafeeq committed:
Fix assignments.clj

shafeeq committed:
Remove tests

shafeeq committed:
Change old methods to params and schema style

shafeeq committed:
Make parsed xml result a vector for consistency

shafeeq committed:
Add some integration tests.
- Register HIT Type
- Create HIT
- Search HIT
- Disable HIT

shafeeq committed:
Bugfix: In hits.clj for get-hits-for-qualification-type

shafeeq committed:
Bugfix: Method name in misc.clj

shafeeq committed:
Remove old code

shafeeq committed:
Bugfix: Change schemas for requester statistics

shafeeq committed:
Add more tests and refactor

shafeeq committed:
Use nomad for config in test

shafeeq committed:
Refactor test and test_helpers

shafeeq committed:
Minor edits

shafeeq committed:
Bugfix: Convert params for SendTestEventNotification

shafeeq committed:
Edit Readme

shafeeq committed:
Bugfix: Wrong operation string and schema in qualifications.clj

shafeeq committed:
Use env varibles for config

shafeeq committed:
Setup travis

shafeeq committed:
Cleanup

shafeeq committed:
Add review policies to schema

shafeeq committed:
Add some unit tests

shafeeq committed:
Change tests namespace

shafeeq committed:
Fix bug in ApproveAssignments schema

shafeeq committed:
Remove redundant assignment method in hits.clj

shafeeq committed:
Update README

shafeeq committed:
Fix typo in method name

shafeeq committed:
Bump version to 0.7.0

shafeeq committed:
Change project url

shafeeq committed:
Update readme with link to getting started guide

shafeeq committed:
Throw exception if access keys are missing

shafeeq committed:
Restore previous config setup after running tests

shafeeq committed:
Fix typos in README

shafeeq committed:
Cleanup
- Move out a util fn from test helpers
- Move test specific files into test/
- Rename some variables for clarity

shafeeq committed:
Add validations for qualification requirement

shafeeq committed:
Initial idea for requester fn

shafeeq committed:
Use single requester fn, remove old functions, opn details in maps

shafeeq committed:
Move qualification validation to qualifications namespace

shafeeq committed:
Move schemas to seperate namespaces

shafeeq committed:
Minor edit + Throw exception on undefined operation key to requester

shafeeq committed:
Update README

shafeeq committed:
Remove special notation for constants

kitallis committed:
Merge pull request #1 from shafeeq/master

PR to track changes

kitallis committed:
Update README.md

kitallis committed:
adding MIT license

kitallis committed:
Update README.md
